src/pks/flow/richards_pk.cc

```diff
@@ -546,6 +555,107 @@ void Richards::Initialize(const Teuchos::Ptr<State>& S)
+      }
+
+      int ncells_per = -1;
+      for (int col=0; col!=ncols; ++col) {
+        auto& col_cells = mesh_->cells_of_column(col);
+        if (ncells_per < 0) ncells_per = col_cells.size();
+        AMANZI_ASSERT(col_cells.size() == ncells_per);
+        auto& col_faces = mesh_->faces_of_column(col);
+        AMANZI_ASSERT(col_faces.size() == col_cells.size()+1);
+        double z_wt = mesh_->face_centroid(col_faces[0])[z_index] + head_wt;
+
+        if (pres_f.get()) (*pres_f)[0][col_faces[0]] = p_atm + rho * g * head_wt;
+        for (int lcv_c=0; lcv_c!=col_cells.size(); ++lcv_c) {
+          AmanziMesh::Entity_ID c = col_cells[lcv_c];
+          AmanziMesh::Entity_ID f = col_faces[lcv_c+1];
+          pres_c[0][c] = p_atm + rho * g * (z_wt - mesh_->cell_centroid(c)[z_index]);
+          if (pres_f.get()) {
+            (*pres_f)[0][f] = p_atm + rho * g * (z_wt - mesh_->face_centroid(f)[z_index]);
+            (*flags)[f] = 1;
+          }
+        }
+      }
+    }
+    if (pres->HasComponent("face")) {
+      // communicate, then deal with horizontal-normal faces
+      pres->ScatterMasterToGhosted("cell");
+      {
+        const Epetra_MultiVector& pres_c = *pres->ViewComponent("cell", false);
+        Epetra_MultiVector& pres_f = *pres->ViewComponent("face", false);
+        for (AmanziMesh::Entity_ID f=0; f!=pres_f.MyLength(); ++f) {
+          if (!(*flags)[f]) {
+            AmanziMesh::Entity_ID_List f_cells;
+            mesh_->face_get_cells(f, AmanziMesh::Parallel_type::ALL, &f_cells);
+            if (f_cells.size() == 1) {
+              // boundary face, use the cell value as the water table is
+              // assumed to parallel the cell structure
+              pres_f[0][f] = pres_c[0][f_cells[0]];
+              (*flags)[f] = 1;
+            } else {
+              AMANZI_ASSERT(f_cells.size() == 2);
+              // interpolate between cells
+              pres_f[0][f] = (pres_c[0][f_cells[0]] + pres_c[0][f_cells[1]]) / 2.;
+              (*flags)[f] = 1;
+            }
+          }
+        }
+      }
+    }
+    S->GetField(key_, name_)->set_initialized();
+  }
+
+  // constant head datum
+  if (plist_->sublist("initial condition").isParameter("hydrostatic water level [m]")) {
+    double z_wt = plist_->sublist("initial condition").get<double>("hydrostatic water level [m]");
+    double rho = plist_->sublist("initial condition").get<double>("hydrostatic water density [kg m^-3]");
+    Teuchos::RCP<const Epetra_Vector> gvec = S->GetConstantVectorData("gravity");
+    int z_index = mesh_->space_dimension() - 1;
+    double g = -(*gvec)[z_index];
+    double p_atm = *S->GetScalarData("atmospheric_pressure");
+
+    Teuchos::RCP<CompositeVector> pres = S->GetFieldData(key_, name_);
+    Epetra_MultiVector& pres_c = *pres->ViewComponent("cell", false);
+    for (int c=0; c!=pres_c.MyLength(); ++c) {
+      pres_c[0][c] = p_atm + rho * g * (z_wt - mesh_->cell_centroid(c)[z_index]);
```

Comment:
Thinking about spinup -- initial water table in the subsurface. So if my datum is the lowest point in the mesh then I get unsaturated zone above the datum, and if my datum is the highest point in the mesh then I get meters of ponded water on the surface. Can this be extended to a case where the top surface (independent of elevation) gets the atmospheric pressure and the entire subsurface is saturated. Then run it for a certain time to get a steady state for water table initialization. Just a thought.

Reply:
this will avoid the initialization from a column

Reply:
Yes, this covers that case.  Note there are two new conditions -- "hydrostatic head [m]" and "hydrostatic water level [m]".  "hydrostatic head" covers the case where you want things relative to the surface as your datum, whereas "hydrostatic water level" covers the case where you want to set things relative to an absolute datum/elevation.  So setting "hydrostatic head [m]" = 0 would do what you are describing.

This was intended to make "initialize from column" die a painful painful death so that we never use it again.  Of course, we may need to use it for temperature anyway...